pyra/hardware.py

```diff
@@ -0,0 +1,424 @@
+    relative_time=[],
+    cpu_system=[],
+    memory_system=[],
+    network_sent=[],
+    network_received=[]
+)
+
+history_length = 120
+
+
+def update_cpu() -> float:
+    """
+    Update dashboard stats for system CPU usage.
+
+    This will append a new value to the ``dash_stats['cpu_system']`` list.
+
+    Returns
+    -------
+    float
+        The current system cpu percentage utilized.
+
+    Examples
+    --------
+    >>> update_cpu()
+    """
+    cpu_percent = min(100, psutil.cpu_percent(interval=None, percpu=False))  # max of 100
+
+    if initialized:
+        dash_stats['cpu_system'].append(cpu_percent)
+
+    return cpu_percent
+
+
+def update_gpu():
+    """
+    Update dashboard stats for system GPU usage.
+
+    This will create new keys for the ``dash_stats`` dictionary if required, and then append a new value to the
+    appropriate list.
+
+    Nvidia GPUs are fully supported. AMD GPUs are currently only supported on Linux.
+
+    Examples
+    --------
+    >>> update_gpu()
+    """
+    global nvidia_gpus
+    nvidia_gpus = GPUtil.getGPUs()  # need to get the GPUs again otherwise the load does not update
+
+    for gpu in nvidia_gpus:
+        name = f'{gpu.name}-{gpu.id}'
+
+        if initialized:
+            try:
+                dash_stats[f'gpu_{name}']
+            except KeyError:
+                dash_stats[f'gpu_{name}'] = []
+            finally:
+                gpu_load = min(100, gpu.load * 100)  # convert decimal to percentage, max of 100
+                dash_stats[f'gpu_{name}'].append(gpu_load)
+
+    # this is untested
+    for amd_index in range(amd_gpus):
+        gpu = pyamdgpuinfo.get_gpu(amd_index)
+
+        name = f'{gpu.name}-{gpu.gpu_id}'
+
+        if initialized:
+            try:
+                dash_stats[f'gpu_{name}']
+            except KeyError:
+                dash_stats[f'gpu_{name}'] = []
+            finally:
+                gpu_load = min(100, gpu.query_load())  # max of 100
+                dash_stats[f'gpu_{name}'].append(gpu_load)
+
+
+def update_memory():
+    """
+    Update dashboard stats for system memory usage.
+
+    This will append a new value to the ``dash_stats['memory_system']`` list.
+
+    Returns
+    -------
+    float
+        The current system memory percentage utilized.
+
+    Examples
+    --------
+    >>> update_memory()
+    """
+    memory_percent = min(100, psutil.virtual_memory().percent)  # max of 100
+
+    if initialized:
+        dash_stats['memory_system'].append(memory_percent)
+
+    return memory_percent
+
+
+def update_network():
+    """
+    Update dashboard stats for system network usage.
+
+    This will append a new values to the ``dash_stats['network_received']`` and ``dash_stats['network_sent']``
+    lists.
+
+    Returns
+    -------
+    tuple
+        A tuple of the received and sent values as a difference since the last update.
+
+    Examples
+    --------
+    >>> update_network()
+    """
+    global initialized
```

Comment:
I don't see `global initialized` elsewhere, why is it here?

Reply:
line 281

Reply:
I'll remove it, probably copied from the update() function